torch/autograd/gradcheck.py

```diff
@@ -183,43 +224,25 @@ def fail_test(msg):
                   'although analytical gradient matches numerical gradient. '
                   f'The tolerance for nondeterminism was {nondet_tol}.')
     failed = not (reentrant and correct_grad_sizes and correct_grad_types)
-    return jacobian, failed
-
-
-def get_analytical_jacobian(inputs, output, nondet_tol, grad_out):
-    diff_input_list = list(iter_tensors(inputs, True))
-    jacobian = make_jacobian(inputs, output.numel())
-    jacobian_reentrant = make_jacobian(inputs, output.numel())
-    grad_output = torch.zeros_like(output, memory_format=torch.legacy_contiguous_format)
-    flat_grad_output = grad_output.view(-1)
-    reentrant = True
-    correct_grad_sizes = True
-    correct_grad_types = True
-
-    for i in range(flat_grad_output.numel()):
-        flat_grad_output.zero_()
-        flat_grad_output[i] = grad_out
-        for jacobian_c in (jacobian, jacobian_reentrant):
-            grads_input = torch.autograd.grad(output, diff_input_list, grad_output,
-                                              retain_graph=True, allow_unused=True)
-            for jacobian_x, d_x, x in zip(jacobian_c, grads_input, diff_input_list):
-                if d_x is not None and d_x.size() != x.size():
-                    correct_grad_sizes = False
-                elif d_x is not None and d_x.dtype != x.dtype:
-                    correct_grad_types = False
-                elif jacobian_x.numel() != 0:
-                    if d_x is None:
-                        jacobian_x[:, i].zero_()
-                    else:
-                        d_x_dense = d_x.to_dense() if not d_x.layout == torch.strided else d_x
-                        assert jacobian_x[:, i].numel() == d_x_dense.numel()
-                        jacobian_x[:, i] = d_x_dense.contiguous().view(-1)
-
-    for jacobian_x, jacobian_reentrant_x in zip(jacobian, jacobian_reentrant):
-        if jacobian_x.numel() != 0 and (jacobian_x - jacobian_reentrant_x).abs().max() > nondet_tol:
-            reentrant = False
-
-    return jacobian, reentrant, correct_grad_sizes, correct_grad_types
+    return jacobians, failed
+
+
+def get_analytical_jacobian(fn, sample_output, grad_out_scale):
+    # Computes Jacobian row-by-row using backward function `fn` = v^T J
+    # NB: we can't combine the rows into a single jacobian tensor because fn(v) for
+    # different v may return tensors with different number of elements
+    grad_out_base = torch.zeros_like(sample_output, memory_format=torch.legacy_contiguous_format)
+    flat_grad_out = grad_out_base.view(-1)
+    # jacobians_rows[i][j] represents the jth row of the ith input
+    jacobians_rows: Dict[int, List[Optional[torch.Tensor]]] = {}
```

Comment:
Why does it need to be a Dict[int, List[Optional[torch.Tensor]]]? Couldn't it be a List[List[Optional[torch.Tensor]]]?

Reply:
Yeah, its probably safe to assume that different grad_outs to backward produce the same number of inputs.